resources/estimate_copy_number.R

```diff
@@ -0,0 +1,150 @@
+  normalizeBins() %>%
+  segmentBins(transformFun="sqrt")
+
+# Generate called read count plots
+# for final CNV call profiles
+copyNumbersCalled <- readCounts %>%
+  
+  # Filter reads based on a) residuals after fitting to a model, 
+  # and b) on amount  of acceptable overlap with 
+  # blacklisted (i.e. problematic) regions/known artifacts
+  applyFilters(residual=TRUE, blacklist=TRUE, chromosomes=NA) %>%
+  
+  # Estimate and correct read counts as a function of GC content and
+  # mappability (removes further noise from e.g. FFPE)
+  estimateCorrection() %>%
+  correctBins() %>%
+  
+  # Normalise and smooth corrected read counts
+  normalizeBins() %>%
+  smoothOutlierBins() %>%
+  
+  # Segment bins: i.e. group similar regions together
+  # in a single "line" or "window" in terms of read count value
+  segmentBins(transformFun="sqrt") %>%
+  normalizeSegmentedBins() %>%
+  callBins()
+
+# Save estimated copy number plots
+#PNG naming needs fixing - "%s_" generates error
+#Error in png(filename = "%s_called_copy_numbers.png", width = 1300, height = 800) : 
+#  invalid 'filename'
+png(filename = paste0(filename, ".copy_number_segmented.png"), width = 1300, height = 800)
+plot(copy_Numbers_segmented)
+dev.off()
+
+png(filename = paste0(filename, ".called_copy_numbers.png"), width = 1300, height = 800)
+plot(copyNumbersCalled)
+dev.off()
+
+# Save tabular data
+
+## Summarised calls - one file per sample
+QDNAseq::exportBins(copyNumbersCalled,
+                    format = "seg", 
+                    type = "calls",
+                    file = paste("%s.summarised_calls_binsize_",binsize,".tsv", sep = ""))
+
+## Raw calls - one file that contains *all* samples together 
+### MAY NOT NEED THIS FUNCTION
+QDNAseq::exportBins(copyNumbersCalled, 
+                    format = "tsv", 
+                    type = "calls",
+                    file = paste(filename, ".all_calls_binsize_",
+                                 binsize, "_merged.raw_calls_all_depths.tsv",
+                                 sep = ""))
+
+################## ACE FUNCTIONS #############################
+
+generate_sample_model <- function(object, QDNAseqobjectsample, filename){
+  # Perform model fitting on a single sample
+  model1 <- singlemodel(object, QDNAseqobjectsample = QDNAseqobjectsample)
+  # Select best fit by examining error lists (cellularity where relative error is lowest)
+  bestfit1 <- model1$minima[tail(which(model1$rerror==min(model1$rerror)), 1)]
+  besterror1 <- min(model1$rerror)
+  lastfit1 <- tail(model1$minima, 1)
+  lasterror1 <- tail(model1$rerror, 1)
+  error_plot <- print(model1$errorplot + ggtitle(paste0("sample ", filename, " - errorlist")) +
+                        theme(plot.title = element_text(hjust = 0.5)))
+  #  Use best fit and best error to select best model for absolute CN plot
+  png(paste0(filename,"_errorplot.png"))
+  ggsave(paste0(filename, "_errorplot.png"), error_plot, limitsize = FALSE)
+  dev.off()
+  
+  return(model1)
+}
```

Comment:
How will you allow the two modes of input of ploidy and cellularity? by previous setting, and by matching the segmentations. Pls. add these options here.